### ISSUE TITLE
IJavascriptCallback sometimes not executing

### ISSUE THREAD

**Issue by ulvesked** [cefsharp/CefSharp#1939]
Using CefSharp.WinForms - v55.0.0 from Nuget

Architecture: x86

Developer machine: Win7
Tested also on Windows 10

Register an object with RegisterJsObject with a method that accepts an IJavascriptCallback parameter. In the method, initialize a timer that fires the callback after 10-15 seconds. The callback is sometimes not executed in the browser.

Expected result: The IJavascriptCallback is executed in the browser when triggered from C#.
Actual result: Occationally the callback sent from C# (using IJavascriptCallback.ExecuteAsync) is not executed in the browser.

debug.log contains these entries:¨
```
[0214/124914:VERBOSE1:dispatcher.cc(534)] Num tracked contexts: 0
[0214/124914:VERBOSE1:script_context.cc(111)] Created context:
  extension id:           (none)
  frame:                  531219DC
  URL:                    
  context_type:           WEB_PAGE
  effective extension id: (none)
  effective context type: WEB_PAGE
[0214/124914:VERBOSE1:script_context.cc(111)] Created context:
  extension id:           (none)
  frame:                  00000000
  URL:                    
  context_type:           UNSPECIFIED
  effective extension id: (none)
  effective context type: UNSPECIFIED
[0214/124914:VERBOSE1:dispatcher.cc(395)] Num tracked contexts: 1
[0214/124914:VERBOSE1:script_context.cc(120)] Destroyed context for extension
  extension id: 
  effective extension id: 
[0214/124914:VERBOSE1:script_context.cc(120)] Destroyed context for extension
  extension id: 
  effective extension id: 
```

Sample application which demonstrates this issue: 
https://github.com/ulvesked/Testing.CefCallbackDebug

The CefCallbackDebug.MinimalExample is a really stripped down sample which rarely shows the issues, but they still occur. Run the application and point your browser to http://localhost:8080 for remote debugging and watch the console for "test" + "success" messages. 

The other project (CefCallbackDebug.csproj) is more like our real application which loads two different webpages and uses more callbacks. The issues seem to occur more frequently in this one. This project also has more friendly debugging.

**Comment by amaitland** [comment]
The `55.0.0` packages were build with `Gitlink`, so you should be able to step into the source yourself.

https://github.com/GitTools/GitLink#gitlink

**Comment by ulvesked** [comment]
When I step into the source, I get to  CefSharpBrowserWrapper::SendProcessMessage (in CefSharpBrowserWrapper.cpp:237)
VS will not let me step into _browser->SendProcessMessage.

`    return _browser->SendProcessMessage(targetProcess, message);`

All the parameters look good, and the message seems to be fine but it is not triggered in the browser context.

**Comment by amaitland** [comment]
You'll have to debug the renderer process then

**Comment by amaitland** [comment]
https://www.chromium.org/developers/how-tos/debugging-on-windows#TOC-Manually-attaching-to-a-child-process

**Comment by kkneip** [comment]
I also encountered the same issue.
Any updates about this topic?

**Comment by amaitland** [comment]
> The CefCallbackDebug.MinimalExample is a really stripped down sample which rarely shows the issues, but they still occur.

Your example calls `Reload` at a random interval and if your callback executes when the page is in the middle of a reload it will fail as there is no valid `V8Context` to execute the javascript. Same problem if you call `Load`.

In the next version you will be able to use `IRenderProcessMessageHandler.OnContextCreated/OnContextReleased` to be notified when the context is available. 
https://github.com/cefsharp/CefSharp/commit/5d459a5d32f4c71d7e3df42b96934f50b44249ac#diff-3bb80054f9bf836a85fc190b42871866R33

I've created #1984 which will outline some improvements to `IJavascriptCallback.CanExecute`, for now you need to take into account a page reload, you can likely use `LoadingStateChanged` to improve your code.

**Comment by ulvesked** [comment]
I have not had the chance to set up a CEF debugging environment. 

> Your example calls Reload at a random interval and if your callback executes when the page is in the middle of a reload it will fail as there is no valid V8Context to execute the javascript. Same problem if you call Load.
I understand this, but the issue is not related to reloads. The issue remains without reloads and in my example it occurs "between" reloads. The context is still available because I can still use the debugger on the same page and the page can use other Cef bound objects.

I made a workaround and use string callbacks instead of IJavascriptCallbacks. These strings are names of global functions and they are executed with `ChromiumWebBrowser.EvaluateScriptAsync`

This works every time, and the IJavascriptCallback does not. However, I like the IJavascriptCallback better, since global functions and "eval"-like invokation are "evil" :).

**Comment by amaitland** [comment]
You only need to download the CefSharp source to debug. 

If Reload and load aren't relevant to reproducing the problem then why are they used in both of your examples? Remove any unnecessary code that's not relevant.

**Comment by ulvesked** [comment]
The issues does not always occur. It only happens on about every 5-10th page load. The reload was implemented for two reasons, to not need to restart the application if the issues could not be reproduced, and partly to be similar to my actual implementation. I will create a new demo this week without reload, but this demo needs to be shut down and restarted several times to reproduce the problem.

**Comment by ulvesked** [comment]
> You only need to download the CefSharp source to debug.

I tried that, and I got into the SendProcessMessage and need to debug the renderer process but I cannot step further down. I thought the SendProcessMessage is CEF internal.

**Comment by amaitland** [comment]
https://github.com/cefsharp/CefSharp/blob/cefsharp/55/CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp#L302

You will need to download the source and attached the debugger as per the Chromium instructions. 

Have you tested a different version?

**Comment by amaitland** [comment]
Your original issue description really should have included information about taking 5-10 page reloads to reproduce.

**Comment by ulvesked** [comment]
Sorry about that. I wrote that "sometimes" and "occationally" the callbacks are not executed. Should have been more clear on that. I will update the original issue.

**Comment by amaitland** [comment]
Where are you checking the response of the callback? Does it return an error message?

**Comment by amaitland** [comment]
I'll reopen this for now, though it's unlikely that I'll look into this further myself (the code is there, anyone is able to debug it quite easily).

Here are my final comments for now:

- Code should be tested against an older version, if there is a bug, it maybe a small `quirk` that's only present in the `2883`. The fact the `JS Callback` code hasn't been touched in many versions and there is not a single report of a problem until version `55`.
- I'm still not convinced there isn't a subtle timing problem in your code, that's just my initial thoughts.
- The example provided ignores the response from the callback, it should be updated to log the error message, which I expect will be `Unable to find callbackWrapper`. 
- It's easy to reproduce a `Unable to find callbackWrapper` error, just execute the callback in the middle of the page `reload/navigation`. My initial comments are based of this known scenario.
- Update the example(s) to wire up `browser.ConsoleMessage` event, and log all the output to the same place, having to manually connect `DevTools` makes it harder than necessary. 
- Strip any unnecessary code out of the more complex example, it's hard to follow the logic flow as is (better code comments would help).
- For debugging purposes the project source should be downloaded and modified to reproduce the problem, this will make debugging the `subprocess` easier. It's very easy to debug the sub process, see https://www.chromium.org/developers/how-tos/debugging-on-windows#TOC-Manually-attaching-to-a-child-process for the basics

**Comment by amaitland** [comment]
For reference I've made callback `Id's` unique per process see https://github.com/cefsharp/CefSharp/commit/910df4877a85d1b88ae8c6b92e86a872612102ec#diff-8dabe4228c63cbb1fba228450b3b7609 this change will be in the `57.0.0` release.

A callback is only valid between `IRenderProcessMessageHandler.OnContextCreated` and `IRenderProcessMessageHandler.OnContextReleased`, executing a callback without a valid context will result in a `Unable to find callbackWrapper` error message. It's currently necessary to handle this behavior yourself. #1984 tracks proposed changes to `IJavascriptCallback.CanExecute`.

`IRenderProcessMessageHandler.OnContextReleased` is part of the `57.0.0` release which is due shortly.
